--- E:\calab\ucas-ca-lab\exp_18_cpu_with_tlb\myCPU\tlb.v
+++ E:\calab\ucas-ca-lab\exp_19_tlb_exc\myCPU\tlb.v
@@ -152,10 +152,12 @@
         for (i = 0; i < TLBNUM; i = i + 1) begin: search
             assign match0[i] = (s0_vppn[18: 9]==tlb_vppn[i][18: 9])
                                 && (tlb_ps4MB[i] || s0_vppn[8:0]==tlb_vppn[i][8:0])
-                                && ((s0_asid==tlb_asid[i]) || tlb_g[i]);
+                                && ((s0_asid==tlb_asid[i]) || tlb_g[i])
+                                && tlb_e[i];
             assign match1[i] = (s1_vppn[18: 9]==tlb_vppn[i][18: 9])
                                 && (tlb_ps4MB[i] || s1_vppn[8:0]==tlb_vppn[i][8:0])
-                                && ((s1_asid==tlb_asid[i]) || tlb_g[i]);
+                                && ((s1_asid==tlb_asid[i]) || tlb_g[i])
+                                && tlb_e[i];
         end
     endgenerate
 
